libbeat/processors/actions/drop_sampling.go

```diff
@@ -0,0 +1,140 @@
+package actions
+
+import (
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type dropSampling struct {
+	rnd       *rand.Rand
+	seed      int64
+	lock      *sync.Mutex	
+	skipped   uint64
+	allowed   uint64
+	rate      float64
+	threshold float64
+	metrics   *PeriodicSamplingMetrics
+}
+
+func init() {
+	processors.RegisterPlugin("drop_sampling", newDropSample)
+}
+
+// accept returns true iff the random number generated is below
+// the configured rate, otherwise false.
+func (d *dropSampling) accept(rate float64) (float64, bool) {
+	threshold := d.rnd.Float64()
+	return threshold, threshold <= rate
+}
+
+func newDropSample(c *common.Config) (processors.Processor, error) {
+	seed := time.Now().UnixNano() // turn into something set by config
+	src := rand.NewSource(seed)
+	rnd := rand.New(src)
+
+	ds := &dropSampling{
+		rnd:           rnd,
+		lock:          &sync.Mutex{},
+		metrics:       StartPeriodicMetrics(),
+	}
+
+	go ds.metrics.run(ds.provideMetrics)
+
+	return ds, nil
+}
+
+func (d *dropSampling) Run(b *beat.Event) (*beat.Event, error) {
+	rate, ok := d.findSampling(b)
+	if !ok {
+		return b, nil
+	}
+
+	threshold, ok := d.accept(rate)
+
+	d.trackMetrics(ok, rate, threshold)
+
+	if ok {
+		return b, nil
+	}
+
+	return nil, nil
+}
+
+func (d *dropSampling) trackMetrics(ok bool, rate, threshold float64) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if ok {
+		d.allowed++
```

Comment:
I went with locking because the same values need to be read when published during periodic output to log.  The idea here was to allow all the values to be gathered under one lock since a single update is required for the four values.  But if we remove threshold and rate, then it's really arbitrary whether to use a lock.